Refactor the instruction tracing subsystem.

qwattash opened the pull request:
Add tracing backend based on the google perfetto tracing framework.
Add the ability to aggregate statistics within qemu and emit them as part of the perfetto backend.
Add support for generic events associated to instructions.
Add better support for trace filters for each log entry.

qwattash committed:
Add initialization and setup code for the perfetto tracing backend.

- Add cheri-perfetto submodule for tracing.
- Plumb the perfetto tracing backend in the build system
- Add initialization code that registers the tracing categories and starts
  the tracing service at startup
- Add skeleton for the perfetto tracing backend glue code

qwattash committed:
Refactor instruction tracing to support events.

- Rename cpu_log_instr_info to cpu_log_entry.

qwattash committed:
Refactor instruction logging: rename trace_format to trace_backend.

qwattash committed:
Implement an event queue for each instruction entry logged.

qwattash committed:
Generalize instruction trace filtering.

The -dfilter qemu option has ad-hoc support within TCG instruction tracing,
this generalizes the trace filtering to an arbitrary per-CPU set of functions
that can filter the trace entry based on any parameter.
Note that for now filters are assumed to be stateless.

qwattash committed:
Refactor TCG log instruction backend into multiple files.

Each trace format is moved to a separate file for implementation.

qwattash committed:
Rename TCG instruction logging backend macros.

qwattash committed:
Add stats TCG instruction logging backend.

This backend is intended to capture statistics from the instructions
instead of dumping the whole instruction stream.
Note that this is disabled by default in the configuration as it adds
a dependency on libboost.

qwattash committed:
Actually emit events.

qwattash committed:
Add magic NOP to log context switch events.

qwattash committed:
Fix stats logging backend output.

qwattash committed:
Improve instruction trace filters support.

- Add command line option to select filters to use.
- Make sure that the filters from command line are applied to the CPUs at
  initialization.
- Add filter to only emit entries containing events.

qwattash committed:
Refactor trace stats helper into a class.

qwattash committed:
Implement a tracing backend based on the perfetto tracing framework.

For each CPU we create 2 tracks: one for control events (e.g. tracing
start/stop) and another for log events on that CPU.
Currently we emit instructions and stats events on the CPU track.

- Replace the stats backend with a tracing category in the perfetto backend
- Add new CLI options to control the perfetto backend behaviour:
  1. --cheri-trace-perfetto-categories comma-separated list of categories to
     enable.
  2. --cheri-trace-perfetto-logfile log file to use for the perfetto binary
     trace. We do not use the common qemu logfile to avoid accidental pollution
     and passing the qemu file descriptor is painful due to having to lock the
     logfile via RCU.
- Add hook at qemu shutdown to make sure we flush the log buffers, if the
  tracing backend supports this.

qwattash committed:
Fix broken legacy CVTrace backend.

qwattash committed:
Fix flushing in instruction logging perfetto backend.

Flush TrackEvent data source when we receive a FLUSH event in the instruction
logging subsystem. This helps making sure that we do not have lost events when
running multiple tests on qemu.

arichardson commented on accel/tcg/log_instr.c:
```diff
@@ -400,36 +408,32 @@ static void emit_text_entry(CPUArchState *env, cpu_log_entry_t *entry)
-                 env_cpu(env)->cpu_index, cpu_get_asid(env), pc);
+    for (i = 0; i < entry->events->len; i++) {
+        event = &g_array_index(entry->events, const log_event_t, i);
+        if (event->id == LOG_EVENT_STATE) {
```
add `else { assert(0 && "unknown event id")' }`?

arichardson commented on include/exec/log_instr.h:
```diff
@@ -180,6 +180,50 @@ struct TCGv_i64_d;
+ */
+typedef enum {
+    LOG_EVENT_STATE = 0,
+    LOG_EVENT_CTX_SWITCH = 1,
```
Unused in this commit?

arichardson commented on target/riscv/op_helper_log_instr.c:
```diff
@@ -52,3 +59,32 @@ void HELPER(riscv_log_instr)(CPURISCVState *env, target_ulong pc,
+        event.ctx_switch.cid = get_gpr_value(env, 13);
+        break;
+    default:
+        return;
```
warn about an invalid event here?

qwattash commented on configure:
```diff
@@ -340,6 +340,7 @@ pvrdma=""
 gprof="no"
 debug_tcg="no"
 tcg_log_instr="yes"
+tcg_log_instr_perfetto="yes"
```
This should probably be off by default as we pull in some of the boost libraries and the perfetto submodule as a dependency.

arichardson commented on stats/meson.build:
```diff
@@ -0,0 +1,12 @@
+trace_stats_ss = ss.source_set()
+trace_stats_ss.add(files('trace_stats.cc'))
+
+boost_dep = dependency('boost', modules: ['iostreams'])
```
I guess most of us have boost installed, but I think this is a new dep?

arichardson replied:
Ah but it's off by default so this really doesn't matter :)

arichardson reviewed (commented):
This generally looks fine to me (only skimmed most of the commits though). Since it's off by default it shouldn't really matter that we pull in boost.

qwattash committed:
Disable the perfetto tracing backend by default.

Write perfetto configuration status when running configure.
Enable perfetto on Linux CI builds.

qwattash committed:
Fix tracking of the current basic-block in perfetto trace stats.

Reset state when tracing is started and stopped.

qwattash committed:
Implement guest-driven context tracking for the perfetto log backend.

The guest OS can use reserved event NOPs to inform the perfetto backend
about context changes (e.g. a context switch or a compartment switch).
The perfetto backend allocates a separate track to each context and
emits events to the active track in each CPU.
Note that the notion of context switch is entirely given by the guest OS
and it is responsible for correctly updating the state on the backend.
The probe effect in the code due to NOPs instructions should be negligible.

qwattash committed:
Modify meaning of the basic-block hit histogram tracing statistic

Record number of instructions hit for each basic-block instead of the
number of hits per basic-block.

qwattash committed:
Fix use of icount in BB histogram

qwattash committed:
Properly propagate the PC for tracing START/STOP events.

qwattash committed:
Add TODO for better tracing context switches in perfetto backend.

qwattash committed:
Fix instruction logging backend context tracing.

- Add extra LOG_EVENT_CTX_OP_SETUP variant of the OP_SWITCH context update
event.
  This is propagated even when tracing is off and can be used to setup the
  context information before tracing starts.
- Fix statistics trace state reset when tracing is paused and when the
  active context changes.
- Add a subclass for QEMU context tracks. This follows the pattern of perfetto
  ThreadTrack and ProcessTrack.
- Add events to RISC-V target.

qwattash committed:
Fix inline function prototype to silence warnings.

qwattash committed:
Update user logging to match the use of QEMU_LOG_INSTR_U.

qwattash committed:
Handle unlikely case for qemu perfetto stats recording.

Avoid creating histogram buckets starting at 0 when pause is called
but no instructions have been executed yet.

qwattash committed:
Refactor the way instruction logging statistics are handled.

 - Use a separate statistics object for each context detected.
 - Rate-limit the dumping of the stats histogram to trace so that we
   do not uselessy generate huge traces.

qwattash committed:
Fix support for the instruction logging EVENT_CTX_UPDATE.

Unbreak handling of LOG_INSTR_EVENT_CTX_OP_SETUP operation so that it
is correctly propagated to the backend.

qwattash committed:
Fix handling of context update events in text instruction logging backend.

qwattash committed:
Minor format and comment fix in instruction logging implemenation.

qwattash committed:
Fix building of ARM target with refactored instruction trace backends.

qwattash committed:
Properly handle perfetto logging backend dependency resolution.

qwattash committed:
Add missing dependency for perfetto tracing backend.

qwattash committed:
Pass instruction logging code through clang-format.

qwattash committed:
Update perfetto instruction logging to match cheri-perfetto histogram messages.

qwattash committed:
Comply with CHERI-architecture reserved RISC-V opcodes for tracing.

qwattash committed:
Introduce support for guest-defined trace markers.

qwattash committed:
Add instruction logging backend that emits a protobuf stream.

This is inspired to the perfetto log backend and has two main use
cases:
 - Provide a portable way to parse full instruction traces, without
 requiring all the perfetto backend dependencies.
 - Detect performance and disk-space issues with the perfetto backend
 serialization mechanism.
Note that this backend only logs the instruction stream and does not
support all the extra data framing, selection and aggregation that
the perfetto backend performs.

qwattash committed:
Add experimental instruction logging backend that generates JSON data.

This should have similar efficiency to the text instruction format
but be more stable and easier to parse. The protobuf backend should
be preferred for parsing.

qwattash committed:
Fix build without the cheri-perfetto submodule.

Only look for cheri-perfetto files if perfetto instruction logging is enabled.

qwattash committed:
Update CHERI perfetto interface.

 - Use the same protobuf descriptor for instructions as the one used
 by the protobuf backend.
 - Use the CHERI context track implementation to embed context
 identifiers in the trace instead of using the Track name.

qwattash committed:
Use perfetto protobuf definition for the protobuf backend.

qwattash committed:
Implement log event to dump register contents to trace.

Introduce target-dependent hooks to produce register dumps in the
instruction logging subsystem.
Add interface functions for target to emit registers in the event,
this keeps log_reginfo_t private.

qwattash committed:
Fix allocation of temporary argument to do_cpu_loglevel_switch.

The function expects an argument structure allocated by g_new.

qwattash committed:
Remove log event variant that bypasses log enable.

After testing this was deemed unnecessary.

qwattash committed:
Implement register dumps for protobuf instruction tracing backend.

qwattash committed:
Retire instruction logging event type for context updates.

Remove the event variant that bypasses log enable. This is useless as
switching on/off tracing around the event is feasible and the event is
rare enough that it is not a performance issue.

qwattash committed:
Implement RISC-V tracing counter NOPs.

Add NOP that guests can use to create and increment/decrement
counters.
The counters are emitted into the perfetto trace stream as normal
perfetto counter tracks.

qwattash committed:
Point cheri-perfetto submodule to the correct repository.

qwattash committed:
Add per-context counters.

These are still not hooked to the guest NOPs.

qwattash committed:
Introduce interval tracks support in the perfetto tracing backend.

Replace the stats implementation with a new one based on the
cheri-perfetto interval tracks.
The stats implementation tracks basic block icount and hits, along
with branches. Now there are multiple categories for each different
statistic instead of a grouped one.

qwattash committed:
Fix instruction opcode dumping to the perfetto message format.

This works around the inability to use bytes fields in perfetto
track event messages.

qwattash committed:
Fix cheri-perfetto new remote URL.

qwattash committed:
Add option to select output file for the protobuf tracing backend.

Commit:
Add an experimental interceptor in perfetto to trace memory load and store

qwattash committed:
Unbreak -icount mode for RISC-V.

qwattash committed:
Drop the instruction tracing backend emit_header hook.

This was essentially unused, and can be replicated during initialization.

Commit:
add gzip compression for interceptor output file

qwattash committed:
Improve perfetto trace session configuration.

qwattash committed:
Introduce support for logging qemu debugging counters with perfetto.

Add two counters to measure instruction count when tracing is enabled/disabled.

qwattash committed:
Disable branch and basic block tracking when unnecessary.

Only maintain the branch and basic block hit histograms when
the corresponding trace categories are enabled.

Commit:
Point dynamrio submodule to CHERI fork

qwattash committed:
Disable tb_flush when switching log level.

This should not be needed because we have flags marking TBs whenever
they contain instruction tracing instrumentation.

qwattash committed:
Fix io_uring_prep_poll_remove call.

This is broken by recent liburing versions.

qwattash committed:
Always use a fixed number of entries in the trace backends list.

qwattash committed:
Use disas_one_str in protobuf backend.

qwattash committed:
Update disasm helpers to emit single instructions as strings.

Add disas_one_strbuf() to disassemble one instruction to an existing
string buffer.
Add disas_one_str() to disassemble one instruction to a dynamically
allocated string.

qwattash committed:
Add explicit file parameter to non-text tracing backends.

- Conditionally compile cvtrace
- This is done because some of these backends emit binary data that
  is logically incompatible with the standard qemu output file.
  The downside is the we need to handle our own locking if we
  want to support multiple CPUs.

qwattash committed:
Cleanup signatures in instruction logging infrastructure.

qwattash committed:
Add the current log level transition in the log state transition event.

This should help to inform log backends for management purposes.

qwattash committed:
Rename per-CPU instruction logging buffer to entry_buffer.

This is better than instr_info as it reflects its use.

qwattash committed:
Remove perfetto debug counters.

qwattash committed:
Restore entry buffer size var.
